### DISCUSSION TITLE
IdleNotifications Custom Sound Selection

### DISCUSSION THREAD

**Discussion by jaybyrrd** [runelite/runelite#18038] [Ideas]
Wanted to start a discussion around Custom Sounds in Idle Notifier Plugin. This came up recently in [discord](https://discord.com/channels/301497432909414422/419891709883973642/1255500349003923550). 

Mostly providing this write-up of thoughts since I am intending to do this work for the plugin and am happy to improve the approach. 

## Where we are now

Many of the features asked for in this other [discussion](https://github.com/runelite/runelite/discussions/17114) have been implemented with exception to the Mouse Tooltip (which is likely worth a continued discussion on). 

Custom Sounds appears to be partially implemented but a lot of the way there. This discussion is going to focus on continuing that work. 

### User Experience

At a high level, the current flow for customizing a Notification is as follows:

1. The user opens Idle Notifier and Enables A Notification
2. The users clicks a setting cog next to the enabled notification 

![image](https://github.com/user-attachments/assets/f592ebad-b5e2-4571-b74f-3db5b9a03f1a)

3. The user is presented a panel to customize the notification which requires that you first check a customize checkbox
![image](https://github.com/user-attachments/assets/81671db0-436b-4b16-b99e-89c5a96f33de)

4. Once clicked, the user sees a menu with a combo box that presents the options `Native`, `Custom`, `Off`
![image](https://github.com/user-attachments/assets/b58524da-ccee-462b-b07b-1a93ef059273)

This custom selection does not actually allow you to set the sound file played. 

### Code Dive

The [`Notification` Class](https://github.com/runelite/runelite/blob/bf335dc399be7a3592ea4a75680341dcbc97d12f/runelite-client/src/main/java/net/runelite/client/config/Notification.java#L42) represents the configurable parts of a notification in the plugin. The field `Notifier.NativeCustomOff sound` is an [enum](https://github.com/runelite/runelite/blob/bf335dc399be7a3592ea4a75680341dcbc97d12f/runelite-client/src/main/java/net/runelite/client/config/Notification.java#L42) that statically describes the options provided in this [JComboBox](https://github.com/runelite/runelite/blob/b9da19600c6df915196a142dfa2d34080067d03e/runelite-client/src/main/java/net/runelite/client/plugins/config/NotificationPanel.java#L302). 

When a user clicks Custom, the notification type becomes Custom, which ultimately calls this [line](https://github.com/runelite/runelite/blob/54f206647da17ae5e1c9944125d719b07bea27df/runelite-client/src/main/java/net/runelite/client/Notifier.java#L226-L228) when playing a notification. 

The method [`playCustomSound`](https://github.com/runelite/runelite/blob/54f206647da17ae5e1c9944125d719b07bea27df/runelite-client/src/main/java/net/runelite/client/Notifier.java#L490-L517) has a block where the NOTIFICATION_FILE is a static literal named [`NOTIFICATION_FILE`](https://github.com/runelite/runelite/blob/54f206647da17ae5e1c9944125d719b07bea27df/runelite-client/src/main/java/net/runelite/client/Notifier.java#L107C2-L107C100). 

## General Improvement Topics

This section describes changes that will likely be needed irresepective of the UX. 

### Notification

The notification itself does not have a property for defining a custom file location.

While the `Notifier.NativeCustomOff` enum is valuable because it has allowed us to inject a "custom" code path, it is limiting the way we get input since it is directly used in the combo box for sound selection. We may want to do something like the enum from the selection box and treat the selection box as a Map<String, Notification>, populating it with a runtime populated list of potential sounds of which `Native` and `Off` will always appear. Custom sounds will be the Filename of every `.wav` file in `RuneLite.RUNELITE_DIR/idlenotifierplugin/sounds`. Selecting a file name will provide a `Notification` with `SoundType` as `CUSTOM` and with the corresponding `soundFileLocation` set. 

### NOTIFICATION_FILE

NOTIFICATION_FILE seems like a sane default, but it is right now used in methods that should probably have the sound parameterized in `Notification`. Doing a general refactor around NOTIFICATION_FILE while maintaining it as a sane default is probably the right call.

While NOTIFICATION_FILE should not be moved to a location like `RuneLite.RUNELITE_DIR/idlenotifierplugin/sounds`, custom sounds should be. Not moving the file itself is because this default could be used across plugins. When soundFileLocation is not provided/does not exist and NativeCustomOff is `CUSTOM`, we can default to our existing `NOTIFICATION_FILE`, so that we preserve existng behavior.

### Combo Box
We are effectively missing sound file selection in the user experience. 

The current drop down name suggests that you should be able to, but defining the selection options in the UX based on the `Notifier.NativeCustomOff` actually means that adding new options requires a code change. In order to provide file selection, either the existing or a new config item needs to be collected from the user for which sound to use. We have three options to do this:

1. Update the Combo Box to be populated by a runtime string list
2. Add a second text field for file path that is enabled/considered only when Custom is selected
3. Add a second combo box for file path

Of these options, I am taking the opinion that the best user experience:
- Does not require manual typing of file names
- Requires the user to set as few of fields as possible.

I am open to discussing this opinion, but I think my work towards this problem will involve making it so that the existing Combo Box has a dynamic runtime list so that users can simply add new `.wav` files to the right directory and select those files in a dropdown list.

### Sound File Types

One thing that is unclear to me is whether this only supports `.wav` files or if there is support for `.mp3` already. Depending on that finding, we may want to add support for `.mp3`.

**Comment by jaybyrrd** [comment]
Capturing discord discussion with Adam [here](https://github.com/runelite/runelite/discussions/18038):

1. The ComboBox feature should be built dynamically rather than being tied to a static enum.
2. Notification sounds will be managed by allowing users to add files to a directory, making the menu dynamic.
3. If a user switches devices and a custom sound file is missing, the system will default to playing a pre-existing sound without overwriting the setting.

**Comment by ShaunDreclin** [comment]
Just poking in to see if there is any activity here, or if not, to show demand for proper custom notification sound support!

I'd really like to be able to set different sounds for different notifications.

**Comment by king-koda** [comment]
yeh having this would be great.
